src/exception.c

```diff
@@ -281,16 +289,97 @@ void __onCriticalException(volatile reg_block_t* regs)
 }
 
 /**
- * @brief Respond to a reset exception
+ * @brief Register a handler that will be called when the user
+ *        presses the RESET button. 
+ * 
+ * The N64 sends an interrupt when the RESET button is pressed,
+ * and then actually resets the console after about ~500ms (but less
+ * on some models, see #RESET_TIME_LENGTH).
+ * 
+ * Registering a handler can be used to perform a clean reset.
+ * Technically, at the hardware level, it is important that the RCP
+ * is completely idle when the reset happens, or it might freeze
+ * and require a power-cycle to unfreeze. This means that any
+ * I/O, audio, video activity must cease before #RESET_TIME_LENGTH
+ * has elapsed.
+ * 
+ * This entry point can be used by the game code to basically
+ * halts itself and stops issuing commands. Libdragon itself will
+ * register handlers to halt internal modules so to provide a basic
+ * good reset experience.
+ * 
+ * Handlers can use #exception_reset_time to read how much has passed
+ * since the RESET button was pressed.
+ * 
+ * @param cb    Callback to invoke when the reset button is pressed.
+ * 
+ * @note  Reset handlers are called under interrupt.
+ * 
  */
-void __onResetException(volatile reg_block_t* regs)
+void register_reset_handler( void (*cb)(void) )
 {
-	exception_t e;
-	
-	if(!__exception_handler) { return; }
+	for (int i=0;i<MAX_RESET_HANDLERS;i++)
+	{		
+		if (!__prenmi_handlers[i])
+		{
+			__prenmi_handlers[i] = cb;
+			return;
+		}
+	}
+	assertf(0, "Too many pre-NMI handlers\n");
+}
 
-	__fetch_regs(&e, EXCEPTION_TYPE_RESET, regs);
-	__exception_handler(&e);
+/** 
+ * @brief Check whether the RESET button was pressed and how long we are into
+ *        the reset process.
+ * 
+ * This function returns how many ticks have elapsed since the user has pressed
+ * the RESET button, or 0 if the user has not pressed it.
```

Comment:
Does the counter start when the button is pressed or released? I remember you provided a test rom at some point but I don't remember the behavior RN. Maybe we can include that small piece of information as well.

Reply:
Done